### PR TITLE
[SPARK-49043][SQL] Fix interpreted codepath group by on map containing collated strings

### PR DESCRIPTION
**Pull request by ilicmarkodb**
### What changes were proposed in this pull request?
Added ordering for PhysicalMapType in `PhysicalDataType.scala`.

### Why are the changes needed?
This feature is needed to compare maps for equality in group-by queries when they contain collated strings. 
It was already functional in the codegen path.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added tests to `CollationSuite.scala`

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Review by stefankandic — COMMENTED** [review]
Nice work! Left some minor comments

File: sql/core/src/test/resources/sql-tests/results/mode.sql.out [hunk 1]
```diff
@@ -182,15 +182,9 @@ struct<mode(col):map<int,string>>
 -- !query
 SELECT mode(col, true) FROM VALUES (map(1, 'a')) AS tab(col)
 -- !query schema
-struct<>
+struct<mode() WITHIN GROUP (ORDER BY col DESC):map<int,string>>
```
[the remark this answers is not in the document]
Comment (ilicmarkodb) [reply sql/core/src/test/resources/sql-tests/results/mode.sql.out:185]: I think that when a query throws an exception, the schema in the golden file appears empty. This seems to be the case for all queries in this golden file. However, I don't believe this necessarily means that the schema is actually empty.

Also, when I run the Spark shell on the master code, I get the following output:
scala> sql("SELECT mode(col, true) FROM VALUES (map(1, 'a')) AS tab(col)")
val res0: org.apache.spark.sql.DataFrame = [mode() WITHIN GROUP (ORDER BY col DESC): map<int,string>]

**Comment by ilicmarkodb** [comment]
@cloud-fan can you please review?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/PhysicalDataType.scala [hunk 2]
```diff
@@ -234,10 +234,72 @@ case object PhysicalLongType extends PhysicalLongType
...
   @transient private[sql] lazy val tag = typeTag[InternalType]
+
+  @transient
+  private[sql] lazy val interpretedOrdering: Ordering[MapData] = new Ordering[MapData] {
+    private[this] val keyOrdering =
+      PhysicalDataType(keyType).ordering.asInstanceOf[Ordering[Any]]
+    private[this] val valuesOrdering =
+      PhysicalDataType(valueType).ordering.asInstanceOf[Ordering[Any]]
+
+    override def compare(left: MapData, right: MapData): Int = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/PhysicalDataType.scala:249]: how do we sort map values in the codegen path?

Comment (ilicmarkodb) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/PhysicalDataType.scala:249]: Same as here:
https://github.com/apache/spark/blob/469ebcf4fe41e4c951f4c061ffc31a2dd5a89fc3/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala#L716

Comment (ilicmarkodb) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/PhysicalDataType.scala:249]: But order by is also not supported in codegen path.

**Comment by ilicmarkodb** [comment]
@MaxGekk can you please review?

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/resources/sql-tests/results/mode.sql.out (hunk 1 above)
[the remark this answers is not in the document]
Comment (stefankandic) [reply sql/core/src/test/resources/sql-tests/results/mode.sql.out:185]: Didn't know about that, thanks!

**Review by stefankandic — APPROVED** [review]

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
@ilicmarkodb You pointed out in the ticket SPARK-49043 the affected version 3.4.3. Should we backport this fix to `branch-3.5` and `branch-3.4`. Please, confirm that the branches suffer from the same issue.

**Pull request closed without merging**